File: interpreter/cheese.py
import sys, threading, os

# define stack, memory, line counter, labels, valid commands, and types
debug = len(sys.argv) > 2
line = 0
finished = False
stack = []
memory = [0 for _ in range(4096)]
labels = {}
cmds = {'jmp':1,'set':2,'add':2,'sub':2,'mul':2,'div':2,'mod':2,'psh':1,'pop':1,'swp':1,'inc':1,'dec':1,'clr':1,'cjp':2,'int':0,'pnt':1}
types = {'x':16,'b':2,'d':10}


# error handler
def error(message, typerr=0):
	err = 'ERROR'
	if typerr:
		err = 'WARN'
	print(f'[{err}]: {message} (Line {line+1})')
	if not typerr:
		input()
		sys.exit()


# command interpreter
def interpret_cmd(tokens):
	global line, stack, memory
	# interpret token values
	cmd = tokens[0]
	tokens = tokens[1:]

	if len(tokens) != cmds[cmd]:
		error('Argument error')

	itokens = []
	for token in tokens:
		if token[0] == '@':
			try:
				itokens.append(labels[token[1:]])
			except:
				error(f'Unknown label "{token[1:]}"')
			continue
		tk = token.replace('%','')
		if tk[0] not in types:
			error(f'Unknown type "{tk[0]}"')
		try:
			value = int(tk[1:],types[tk[0]])
		except:
			error(f'Integer parse error "{tk}"')
		if token.startswith('%'):
			try:
				value = memory[value]
				if token.startswith('%%'):
					value = memory[value]
			except:
				error(f'Memory address "{value}" does not exist')
		itokens.append(value)
	tokens = itokens


	# interpret the command
	try:
		if cmd == 'jmp':
			line = tokens[0]-1
			return
		if cmd == 'set':
			memory[tokens[0]] = tokens[1]
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'add':
			memory[tokens[0]] += tokens[1]
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'sub':
			memory[tokens[0]] -= tokens[1]
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'mul':
			memory[tokens[0]] *= tokens[1]
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'div':
			try:
				memory[tokens[0]] //= tokens[1]
			except:
				error('Divide by zero')
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'mod':
			try:
				memory[tokens[0]] %= tokens[1]
			except:
				error('Divide by zero')
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'psh':
			stack.append(tokens[0])
			if len(stack) > 256:
				error('Stack overflow')
			return

		if cmd == 'pop':
			a = memory[tokens[0]]
			try:
				memory[tokens[0]] = stack[-1]
				stack = stack[:-1]
			except:
				error('Stack is empty')
			return

		if cmd == 'swp':
			a = memory[tokens[0]]
			try:
				memory[tokens[0]], stack[-1] = stack[-1], memory[tokens[0]]
			except:
				error('Stack is empty')
			return

		if cmd == 'inc':
			memory[tokens[0]] += 1
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'dec':
			memory[tokens[0]] -= 1
			memory[tokens[0]] %= 4294967296
			return

		if cmd == 'clr':
			memory[tokens[0]] = 0
			return

		if cmd == 'cjp':
			if not tokens[1]:
				line = tokens[0]-1
			return


		if cmd == 'int':
			intret = ''
			for value in memory[3840:]:
				if value == 0:
					break
				try:
					intret += chr(value)
				except:
					error(f'Invalid character "{value}"')
			print(intret)

		if cmd == 'pnt':
			print(tokens[0])
			return

	except:
		error(f'Memory address "{tokens[0]}" does not exist')


# user input loop
def uiloop():
	global memory
	while True:
		ui = input()
		for index, char in enumerate(ui):
			if index > 255:
				error('Input exceeds address xEFF')
			memory[index+3584] = ord(char)


# read main file and remove newline characters
with open(sys.argv[1], 'r') as main:
	main = [line.replace('\n','') for line in main.readlines()]


# find labels and put them in the label dictionary
for ln, contents in enumerate(main):
	if contents.startswith('@'):
		label = contents[1:].split(' ')[0]
		if label in labels.keys():
			error(f'Label "{label}" overrides existing label',1)
		labels[label] = ln+1

# start the ui loop in a new thread
threading.Thread(target=uiloop,daemon=True).start()


# main interpreter
while line < len(main):
	lc = main[line]
	if debug: print(lc)
	if lc == '' or lc.startswith('@') or lc.startswith('~'):
		line+=1
		memory[3568] = line
		continue
	tokens = lc.split(' ')
	if tokens[0] not in cmds:
		error(f'Unknown command "{tokens[0]}"')
	interpret_cmd(tokens)
	line += 1
	memory[3568] = line


if debug: print(stack, memory)
print('[SUCCESS]: Program finished!')
input()


#commands:

# jmp <arg> : jumps to line
# cjp <arg> <arg> : jumps to first argument if second argument is zero
# set <arg> <arg> : sets a registers value to second argument
# add <arg> <arg> : adds a registers value by second argument
# sub <arg> <arg> : subtracts a registers value by second argument
# mul <arg> <arg> : multiplies a registers value by second argument
# div <arg> <arg> : divides a registers value by second argument and discards remainder
# mod <arg> <arg> : adds a registers value by second argument
# psh <arg> : pushes register value onto stack
# pop <arg> : pops top value of stack into register
# swp <arg> : swaps registers value with top of stack
# inc <arg> : increments register
# dec <arg> : decrements register
# clr <arg> : resets register to zero
# int : interrupts program to display contents of xF00-xFFF
# ~comment

#types:

# d<dec> dec
# x<hex> hex
# b<bin> bin
# %<number> value of register
# <number> numerical value
# @<label> label

#syntax:

# <cmd> <arg> <arg>
# @<label>
# ~<comment>

# memory:

#x000-xDEF : program memory
#xDF0-xDFF : important program statistics
#xE00-xEFF : user input
#xF00-xFFF : program output (printed with the int command)

File: interpreter/cheese.cs
using System;
using System.IO;
using System.Linq;
using System.Threading;
using System.Globalization;
using System.Collections.Generic;

class Interpreter {
	public int index = 0;
	public int[] memory = new int[4096];
	public List<int> stack = new List<int>();
	public Dictionary<string, int> cmds = new Dictionary<string, int>() {{"jmp",1},{"cjp",2},{"set",2},{"add",2},{"sub",2},{"mul",2},{"div",2},{"mod",2},{"psh",1},{"pop",1},{"swp",1},{"inc",1},{"dec",1},{"clr",1},{"int", 0}};
	public Dictionary<char, int> types = new Dictionary<char, int>() {{'x',16},{'b',2},{'d',10}};

	public static void Main(string[] args) {
		if (args.Length == 0 || !File.Exists(args[0])) {
			Environment.Exit(0);
		}
		string[] lines = File.ReadAllLines(args[0]);
		new Interpreter().Start(lines);
	}

	public void Error(string err) {
		Console.WriteLine("[ERROR]: {0} (Line {1})", err, index);
		Environment.Exit(0);
	}

	public void Warn(string warn) {
		Console.WriteLine("[WARN]: {0} (Line {1})", warn, index);
	}

	public void Start(string[] lines) {
		for (int i = 0; i < 4096; i++) {
			memory[i] = 0;
		}
		Thread ioloop = new Thread(IOLoop);
		ioloop.IsBackground = true;
		ioloop.Start();
		Dictionary<string, int> labels = FindLabels(lines);
		List<List<string>> tokens = SplitTokens(lines);
		Interpret(tokens, labels);
	}

	public List<List<string>> SplitTokens(string[] lines) {
		List<List<string>> tokens = new List<List<string>>();
		foreach (string line in lines) {
			List<string> toadd = new List<string>();
			if (line == "" || line[0] == '~' || line[0] == '@') {
				toadd.Add(line);
				tokens.Add(toadd);
				continue;
			}
			toadd = line.Split(' ').ToList();
			tokens.Add(toadd);
		}
		return tokens;
	}

	public void IOLoop() {
		while (true) {
			string input = Console.ReadLine();
			if (input.Length > 255) {
				Error("Input exceeds address xEFF");
			}
			for (int i = 0; i < input.Length; i++) {
				memory[i+3584] = (int)input[i];
			}
		}
	}

	public Dictionary<string, int> FindLabels(string[] lines) {
		int lIndex = 0;
		Dictionary<string, int> labels = new Dictionary<string, int>();
		foreach (string lc in lines) {
			if (lc == "") {
				lIndex++;
				continue;
			}
			if (lc[0] == '@') {
				string label = lc.Substring(1);
				if (lc.Contains(" ")) {
					Error("Labels cannot contain spaces");
				}
				if (labels.Keys.Contains(label)) {
					Warn("Label \""+label+"\" overrides existing label");
					labels[label] = lIndex;
				} else {
					labels.Add(label, lIndex);
				}
			}
			lIndex++;
		}
		return labels;
	}

	public List<int> EvalTokens(List<string> tokens, Dictionary<string, int> labels) {
		List<int> rettokens = new List<int>();
		foreach (string token in tokens) {
			if (token[0] == '@') {
				string label = token.Substring(1);
				if (!labels.ContainsKey(label)) {
					Error("Unknown label \""+label+"\"");
				}
				rettokens.Add(labels[label]);
				continue;
			}
			int perC = token.Count(s => s == '%');
			char type = token[perC];
			string val = token.Substring(perC+1);
			int cval = 0;
			if (!types.ContainsKey(type)) {
				Error("Unknown type \""+type+"\"");
			}
			try {
				cval = Convert.ToInt32(val,types[type]);
			} catch {
				Error("Integer parse error \""+token.Substring(perC)+"\"");
			}
			for (int i = 0; i < perC; i++) {
				try {
					cval = memory[cval]; 
				} catch {
					Error("Memory address \""+cval+"\" does not exist");
				}
			}
			rettokens.Add(cval);
		}
		return rettokens;
	}

	public void EvalCommand(string cmd, List<int> tokens) {
		if (cmd == "jmp") {
			index = tokens[0]-1;
		}
		if (cmd == "cjp") {
			if (tokens[1] == 0) {
				index = tokens[0]-1;
			}
		}
		if (cmd == "set") {
			memory[tokens[0]] = tokens[1];
		}
		if (cmd == "add") {
			memory[tokens[0]] += tokens[1];
		}
		if (cmd == "sub") {
			memory[tokens[0]] -= tokens[1];
		}
		if (cmd == "mul") {
			memory[tokens[0]] *= tokens[1];
		}
		if (cmd == "div") {
			try {
				memory[tokens[0]] = (int)memory[tokens[0]]/tokens[1];
			} catch {
				Error("Divide by zero");
			}
		}
		if (cmd == "mod") {
			try {
				memory[tokens[0]] %= tokens[1];
			} catch {
				Error("Divide by zero");
			}
		}

		if (cmd == "psh") {
			stack.Add(tokens[0]);
			if (stack.Count > 256) {
				Error("Stack overflow");
			}
		}
		if (cmd == "pop") {
			try {
				memory[tokens[0]] = stack[stack.Count-1];
				stack = stack.GetRange(0,stack.Count-1);
			} catch {
				Error("Stack is empty");
			}
		}
		if (cmd == "swp") {
			try {
				int tmp = stack[stack.Count-1];
				stack[stack.Count-1] = memory[tokens[0]];
				memory[tokens[0]] = tmp;
			} catch {
				Error("Stack is empty");
			}
		}
		if (cmd == "inc") {
			memory[tokens[0]] += 1;
		}
		if (cmd == "dec") {
			memory[tokens[0]] -= 1;
		}
		if (cmd == "clr") {
			memory[tokens[0]] = 0;
		}
		if (cmd == "int") {
			string intret = "";
			foreach (int val in new ArraySegment<int>(memory, 3840, 256)) {
				if (val == 0) {
					break;
				}
				try {
					intret += (char)val;
				}
				catch {
					Error("Invalid character \""+val+"\"");
				}
			}
			Console.WriteLine(intret);
			return;
		}
		memory[tokens[0]] = (int)(memory[tokens[0]]%4294967296);
	}

	public void Interpret(List<List<string>> lines, Dictionary<string, int> labels) {
		while (index < lines.Count) {
			List<string> tokens = lines[index];
			if (tokens[0] == "" || tokens[0][0] == '~' || tokens[0][0] == '@') {
				index++;
				continue;
			}
			string cmd = tokens[0];
			tokens = tokens.GetRange(1, tokens.Count-1);
			if (!cmds.ContainsKey(cmd)) {
				Error("Unknown command \""+cmd+"\"");
			}
			if (tokens.Count != cmds[cmd]) {
				Error("Argument error");
			}
			List<int> itokens = EvalTokens(tokens, labels);
			EvalCommand(cmd, itokens);
			index++;
		}
	}
}

//commands:

// jmp <arg> : jumps to line
// cjp <arg> <arg> : jumps to first argument if second argument is zero
// set <arg> <arg> : sets a registers value to second argument
// add <arg> <arg> : adds a registers value by second argument
// sub <arg> <arg> : subtracts a registers value by second argument
// mul <arg> <arg> : multiplies a registers value by second argument
// div <arg> <arg> : divides a registers value by second argument and discards remainder
// mod <arg> <arg> : adds a registers value by second argument
// psh <arg> : pushes register value onto stack
// pop <arg> : pops top value of stack into register
// swp <arg> : swaps registers value with top of stack
// inc <arg> : increments register
// dec <arg> : decrements register
// clr <arg> : resets register to zero
// int : interrupts program to display contents of xF00-xFFF
// ~comment

//types:

// d<dec> dec
// x<hex> hex
// b<bin> bin
// %<number> value of register
// <number> numerical value
// @<label> label

//syntax:

// <cmd> <arg> <arg>
// @<label>
// ~<comment>

//memory:

// x000-xDEF : program memory
// xDF0-xDFF : important program statistics
// xE00-xEFF : user input
// xF00-xFFF : program output (printed with the int command)